### ISSUE TITLE
Exception:`IndexOutOfRangeException` with grain request between silo hosts

### ISSUE THREAD

**Issue by wzd24** [dotnet/orleans#7665]
Hey all, I am having a problem with a grain request will throw `IndexOutOfRangeException`.

I am using Orleans 3.5.1

I have attached a log. You can see (near the bottom) at [17:44:36 WRN] is the exception, then scrolling up from there you can see what lead up to it.
[warning2022033017 (1).log](https://github.com/dotnet/orleans/files/8381699/warning2022033017.1.log)

Here is the exception I am seeing.
```
[17:44:36 WRN]Orleans.Networking Exception reading message Response S172.20.1.29:11111:386324274*grn/Sailina.Tang.Activities.ActivityItemCompositeGrain/367+1002668@620937ae->S172.20.1.30:11111:386324402*cli/1f83e08f@4df1e4a0 #58637 from remote endpoint 172.20.1.29:11111 to local endpoint 172.20.1.30:46030
System.IndexOutOfRangeException: Index was outside the bounds of the array.
   at Orleans.Serialization.SerializationManager.ResolveTypeName(String typeName)
   at Orleans.Serialization.SerializationManager.ResolveTypeName(String typeName)
   at Orleans.Serialization.BinaryTokenStreamReaderExtensinons.ReadSpecifiedTypeHeader[TReader](TReader this, SerializationManager serializationManager)
   at Orleans.Serialization.BinaryTokenStreamReaderExtensinons.ReadFullTypeHeader[TReader](TReader this, SerializationManager serializationManager, Type expected)
   at Orleans.Serialization.BinaryTokenStreamReaderExtensinons.ReadSpecifiedTypeHeader[TReader](TReader this, SerializationManager serializationManager)
   at Orleans.Serialization.SerializationManager.DeserializeInner[TContext,TReader](SerializationManager sm, Type expected, TContext context, TReader reader)
   at <>f__AnonymousType1`1Deserializer(Type , IDeserializationContext )
   at Orleans.Serialization.SerializationManager.DeserializeInner[TContext,TReader](SerializationManager sm, Type expected, TContext context, TReader reader)
   at Orleans.Serialization.BuiltInTypes.DeserializeValueTuple3[T1,T2,T3](Type expected, IDeserializationContext context)
   at Orleans.Serialization.SerializationManager.DeserializeInner[TContext,TReader](SerializationManager sm, Type expected, TContext context, TReader reader)
   at Orleans.Serialization.BuiltInTypes.DeserializeOrleansResponse(Type expected, IDeserializationContext context)
   at Orleans.Serialization.SerializationManager.DeserializeInner[TContext,TReader](SerializationManager sm, Type expected, TContext context, TReader reader)
   at Orleans.Runtime.Messaging.MessageSerializer.OrleansSerializer`1.Deserialize(ReadOnlySequence`1 input, T& value)
   at Orleans.Runtime.Messaging.MessageSerializer.TryRead(ReadOnlySequence`1& input, Message& message)
   at Orleans.Runtime.Messaging.Connection.ProcessIncoming()
```

below is my code:

```
//Grian class code
      public async Task<(int errorCode,  object obj)> GetCompositeInfo()
      {
          
          var obj = new
          {
              NumLog = _state.State.GetLog(),
          };
          return (0, obj);
      }
//State class code.
        public object GetLog() 
        {
            var numLog = NumLog?.Select(c => new { Id = c.Key, Num = c.Value }).ToList();
            return numLog;
        }
```

**Comment by wzd24** [comment]
I think the name of the compiler-generated anonymous type is the key factor in throwing an exception. Such as: <>f__AnonymousType154<<Newcs>j__TPar> 

```
        internal Type ResolveTypeName(string typeName)
        {
            if (types.TryGetValue(typeName, out var result))
            {
                return result;
            }

            // Attempt to find the qualified type name from the key and resolve the type from the name.
            if (this.typeKeysToQualifiedNames.TryGetValue(typeName, out var fullyQualifiedName) && this.typeResolver.TryResolveType(fullyQualifiedName, out result))
            {
                this.types[typeName] = result;
                return result;
            }

            if (typeName[typeName.Length - 1] == ']')
            {
                // It's an array type declaration: elementType[,,,]
                var j = typeName.LastIndexOf('[');
                // The rank of the array will be the length of the string, minus the index of the [, minus 1; it's the number of commas between the [ and the ]
                var rank = typeName.Length - j - 1;
                var baseName = typeName.Substring(0, j);
                var baseType = ResolveTypeName(baseName);
                return rank == 1 ? baseType.MakeArrayType() : baseType.MakeArrayType(rank);
            }

            var i = typeName.IndexOf('<');
            if (i >= 0)
            {
                // It's a generic type, definitionType<arg1,arg2,arg3,...>
                var baseName = typeName.Substring(0, i) + "'";
                var typeArgs = new List<Type>();
                i++; // Skip the <
                while (i < typeName.Length - 1)
                {
                    // Get the next type argument, watching for matching angle brackets
                    int n = i;
                    int nestingDepth = 0;
                    while (n < typeName.Length - 1)
                    {
                        if (typeName[n] == '<')
                        {
                            nestingDepth++;
                        }
                        else if (typeName[n] == '>')
                        {
                            if (nestingDepth == 0)
                                break;

                            nestingDepth--;
                        }
                        else if (typeName[n] == ',')
                        {
                            if (nestingDepth == 0)
                                break;
                        }
                        n++;
                    }
                    typeArgs.Add(ResolveTypeName(typeName.Substring(i, n - i))); **//look here.**
                    i = n + 1;
                }
                var baseType = ResolveTypeName(baseName + typeArgs.Count);
                return baseType.MakeGenericType(typeArgs.ToArray<Type>());
            }

            throw new TypeAccessException("Type string \"" + typeName + "\" cannot be resolved.");
        }
```

**Comment by ReubenBond** [comment]
Sounds likely, @wzd24. Are you putting anonymous types into grain state?

**Comment by wzd24** [comment]
@ReubenBond No, but I use anonymous types heavily for message definitions between grains.
List this:
```
 public async Task<(int errorCode,  object obj)> GetCompositeInfo()
 {
     
     var obj = new
     {
         NumLog = _state.State.GetLog(),
     };
     return (0, obj);
 }
```

**Comment by ReubenBond** [comment]
That's not supported - please use named types only

**Comment by wzd24** [comment]
Tks,  this is a bad news about us. Do you have any plans to solve this problem? I believe this is helpful for many business scenarios.

**Comment by wzd24** [comment]
Hey all, I reproduced the bug and located the cause of the issue. This bug is triggered if the data structure contains a generic collection of anonymous types. Such as: List<'a>, HashSet<'a>, ToDictionary<int,'a>... 
Samp code:
```

var list = _data.Items.Select(i => new { i.Id, i.Title }).ToList() // Will throw exception. list is List<'a>
var list = _data.Items.Select(i => new { i.Id, i.Title }).ToHashSet() // Will throw exception.
var list = _data.Items.Select(i => new { i.Id, i.Title }).ToDictionary(i=>i.Id) // Will throw exception.

var list = _data.Items.Select(i => new { i.Id, i.Title }).ToList<object>(); //It's working. list is List<object>.

var list = _data.Items.Select(i => new { i.Id, i.Title }).ToList<dynamic>(); //It's working. list is List<dynamic>.

var list = _data.Items.Select(i => new Item{ i.Id, i.Title }).ToList(); //It's working. list is List<Item>.

var list = new List<object>();
_data.Items.Foreach(i=>list.Add(i)); //It's working.
```
Hope it helps you to solve the problem.

**Comment by ReubenBond** [comment]
Fixed in 7.0
